Add hash_append and use it everywhere

jwnimmer-tri opened the pull request:
Relates #6937.  Closes #7190.

This provides an implementation of object hashing (for use with the `std::unordered_foo` containers) based on Howard Hinnant's excellent work in [N3980](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3980.html), proposed for standardization.

The corresponding styleguide change is RobotLocomotion/styleguide#3.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7357)
<!-- Reviewable:end -->

jwnimmer-tri commented:
+(status: do not merge)

At the moment, this PR is intended as a design proposal with code on the table.  If it passes a first pass of design comments, I plan to refine it into a final implementation, which would include adding appropriate test coverage, as well as a concurrent PR to `styleguide` that allows `std::hash` when coordinated with the new `drake/common/hash.h` APIs.

soonho-tri commented:
+@soonho-tri to put it on my queue.

---

Review status: 0 of 27 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-KxYOoCZHa2z2MPe_1hI:b-13irxn)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Checkpoint.

---

Reviewed 19 of 27 files at r1.
Review status: 19 of 27 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/common/hash.h, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiB5-OG1Td1k3yHCZP:-KxiB5-OG1Td1k3yHCZQ:bf06uu9) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/hash.h#L72)):*
> ```objc
>   using drake::hash_append;
>   hasher(item.data(), item.size());
>   hash_append(hasher, item.size());
> ```

Could you explain why we need to consider this?

---

*[drake/common/hash.h, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiBlXhZMaRl6BV_spV:-KxiBlXhZMaRl6BV_spW:b-w2ecyp) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/hash.h#L95)):*
> ```objc
>     hash_append(hasher, *iter);
>   }
>   hash_append(hasher, count);
> ```

A similar question, why do we need this?

---

*[drake/common/hash.h, line 190 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiDiJS5vuwqDecjOQv:-KxiDiJS5vuwqDecjOQw:bvj0ai9) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/hash.h#L190)):*
> ```objc
> 
>  private:
>   Func func_;
> ```

FYI, have `const Func func_;`?

---

*[drake/common/symbolic_expression.cc, line 83 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiIZ8NE2y0URf_r4ta:-KxiIZ8ONlg1XJW2z_RB:bgmdeyu) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_expression.cc#L83)):*
> ```cpp
> void Expression::HashAppend(DelegatingHasher* hasher) const {
>   using drake::hash_append;
>   using Primitive = std::underlying_type<ExpressionKind>::type;
> ```

I think the following should be enough:
```c++
  hash_append(*hasher, get_kind());
```
thanks to the specialization of `hash_append` for enums.

---

*[drake/common/symbolic_expression.h, line 297 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiHX5s5Ir7ozN8JpDF:-KxiHX5s5Ir7ozN8JpDG:b-argq55) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_expression.h#L297)):*
> ```objc
>       // NOLINTNEXTLINE(runtime/references)
>       HashAlgorithm& hasher, const Expression& item) noexcept {
>     DelegatingHasher delegating_hasher(std::bind(
> ```

BTW, how about the following?

```c++
    DelegatingHasher delegating_hasher(
        [&hasher](const void* data, const size_t length) {
          return hasher(data, length);
        });
```

---

*[drake/common/symbolic_expression_cell.cc, line 294 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiJxECKxFDAnft0lVs:-KxiJxECKxFDAnft0lVt:b-odzsg0) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_expression_cell.cc#L294)):*
> ```cpp
> 
> void ExpressionVar::HashAppendDetail(DelegatingHasher* hasher) const {
>   using drake::hash_append;
> ```

META, missing `DRAKE_ASSERT(hasher);`?  (or you want to remove the ones above?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-KxiKLFnq0ek6NiurwtO:by96gdd)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/hash.h, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiB5-OG1Td1k3yHCZP:-KxiL6EPWtxG2i54Bz21:b-wazu38) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/hash.h#L72)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Could you explain why we need to consider this?
</blockquote></details>

https://htmlpreview.github.io/?https://github.com/HowardHinnant/papers/blob/master/hashing.html#hash_append_vector

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
@maddog-tri , could you check the files under `maliput` directory? (+ `common/hash.h`)

---

Reviewed 5 of 27 files at r1.
Review status: 24 of 27 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/common/symbolic_formula.cc, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiL4tN-JiEsg6Xk7AL:-KxiL4tO2tyFkEgigpQm:b-9xi04g) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_formula.cc#L40)):*
> ```cpp
>   using drake::hash_append;
>   using Primitive = std::underlying_type<FormulaKind>::type;
>   hash_append(*hasher, static_cast<Primitive>(get_kind()));
> ```

`hash_append(*hasher, get_kind());` is enough?

---

*[drake/common/symbolic_formula.h, line 196 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiKpLkQZNMrJTuaLfu:-KxiKpLkQZNMrJTuaLfv:bt0urgp) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_formula.h#L196)):*
> ```objc
>       // NOLINTNEXTLINE(runtime/references)
>       HashAlgorithm& hasher, const Formula& item) noexcept {
>     DelegatingHasher delegating_hasher(std::bind(
> ```

Similarly, the following has the same semantics, but I think it's easier to understand:
```c++
    DelegatingHasher delegating_hasher(
        [&hasher](const void* data, const size_t length) {
          return hasher(data, length);
        });
```

---

*[drake/common/symbolic_formula_cell.h, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiLL69qBIdwFb74LM7:-KxiLL69qBIdwFb74LM8:bksezyn) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_formula_cell.h#L37)):*
> ```objc
>   /** Returns kind of formula. */
>   FormulaKind get_kind() const { return kind_; }
>   /** Sends all hash-relevant bytes for this ExpressionCell type into the given
> ```

BTW, `ExpressionCell` -> `FormulaCell`

---

*[drake/common/symbolic_formula_cell.h, line 39 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiLOIi0Zk_9Nf80d57:-KxiLOIi0Zk_9Nf80d58:b-o9ujsh) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_formula_cell.h#L39)):*
> ```objc
>   /** Sends all hash-relevant bytes for this ExpressionCell type into the given
>    * hasher, per the @ref hash_append concept -- except for get_kind(), because
>    * Expression already sends that.
> ```

BTW, `Expression` -> `Formula`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-KxiMDugGRquH79b07Bc:bc8himh)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 17 of 27 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/common/hash.h, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiB5-OG1Td1k3yHCZP:-KxiMRa_qYXlq_4_O770:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/hash.h#L72)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

https://htmlpreview.github.io/?https://github.com/HowardHinnant/papers/blob/master/hashing.html#hash_append_vector
</blockquote></details>

Done.

---

*[drake/common/hash.h, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiBlXhZMaRl6BV_spV:-KxiMRw0OjJzLrcHh-2r:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/hash.h#L95)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

A similar question, why do we need this?
</blockquote></details>

Done.

---

*[drake/common/hash.h, line 190 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiDiJS5vuwqDecjOQv:-KxiMU-e39v8jcQU4_jm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/hash.h#L190)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, have `const Func func_;`?
</blockquote></details>

Done.

---

*[drake/common/symbolic_expression.cc, line 83 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiIZ8NE2y0URf_r4ta:-KxiNMWuAGooN1Dntn3w:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_expression.cc#L83)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I think the following should be enough:
```c++
  hash_append(*hasher, get_kind());
```
thanks to the specialization of `hash_append` for enums.
</blockquote></details>

Done.

---

*[drake/common/symbolic_expression.h, line 297 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiHX5s5Ir7ozN8JpDF:-KxiMi8nsRvpkicb47XI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_expression.h#L297)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, how about the following?

```c++
    DelegatingHasher delegating_hasher(
        [&hasher](const void* data, const size_t length) {
          return hasher(data, length);
        });
```
</blockquote></details>

Done.

---

*[drake/common/symbolic_expression_cell.cc, line 294 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiJxECKxFDAnft0lVs:-KxiMuOWPu8Ji11grPaY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_expression_cell.cc#L294)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

META, missing `DRAKE_ASSERT(hasher);`?  (or you want to remove the ones above?)
</blockquote></details>

Done.

---

*[drake/common/symbolic_formula.cc, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiL4tN-JiEsg6Xk7AL:-KxiNHn44FWkSi1tXtfX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_formula.cc#L40)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

`hash_append(*hasher, get_kind());` is enough?
</blockquote></details>

Done.

---

*[drake/common/symbolic_formula.h, line 196 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiKpLkQZNMrJTuaLfu:-KxiMwlcaFFSvvuH8OeB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_formula.h#L196)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Similarly, the following has the same semantics, but I think it's easier to understand:
```c++
    DelegatingHasher delegating_hasher(
        [&hasher](const void* data, const size_t length) {
          return hasher(data, length);
        });
```
</blockquote></details>

Done.

---

*[drake/common/symbolic_formula_cell.h, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiLL69qBIdwFb74LM7:-KxiN013cC7RohL1bxCP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_formula_cell.h#L37)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `ExpressionCell` -> `FormulaCell`
</blockquote></details>

Done.

---

*[drake/common/symbolic_formula_cell.h, line 39 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiLOIi0Zk_9Nf80d57:-KxiN1JaFmMKpZGXcTeK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0741ed9653e5bb70166d54a1ba1a89304848f5b/drake/common/symbolic_formula_cell.h#L39)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `Expression` -> `Formula`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 3 of 27 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/symbolic_expression_cell.cc, line 359 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiPnXXVx5ncSaAxVpe:-KxiPnXXVx5ncSaAxVpf:b-mos2e6) ([raw file](https://github.com/robotlocomotion/drake/blob/399d6cb2fe3d9e7299fc6be423152fcb5fd61371/drake/common/symbolic_expression_cell.cc#L359)):*
> ```cpp
> void ExpressionConstant::HashAppendDetail(DelegatingHasher* hasher) const {
>   using drake::hash_append;
>   hash_append(*hasher, v_);
> ```

FYI, this one and the following `*::HashAppendDetail` do not have `DRAKE_ASSERT(hasher);`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-KxiQ9vZy5KmtIOGfEvQ:b-uwkowd)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/symbolic_expression_cell.cc, line 359 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiPnXXVx5ncSaAxVpe:-KxiQm2dF3-xn6vF4ozQ:b-rd3hk4) ([raw file](https://github.com/robotlocomotion/drake/blob/399d6cb2fe3d9e7299fc6be423152fcb5fd61371/drake/common/symbolic_expression_cell.cc#L359)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, this one and the following `*::HashAppendDetail` do not have `DRAKE_ASSERT(hasher);`.
</blockquote></details>

OK My thought was they are private methods, whose sole call site (each) obviously has no `nullptr`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/symbolic_expression_cell.cc, line 359 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxiPnXXVx5ncSaAxVpe:-KxiRFgTcA5BSWsfXx7S:b5pbxgp) ([raw file](https://github.com/robotlocomotion/drake/blob/399d6cb2fe3d9e7299fc6be423152fcb5fd61371/drake/common/symbolic_expression_cell.cc#L359)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

OK My thought was they are private methods, whose sole call site (each) obviously has no `nullptr`.
</blockquote></details>

I see. OK.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
I think it's all good now. It could be better if we provide a variadic version of `hash_append` (as discussed in the paper). But it can be added later (given that this PR is a design PR).

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-KxiRx2ppCrcOvVD4KRR:b-4a1qtd)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 18 of 27 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/common/hash.h, line 82 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxnH5dU9mDoveBFzoie:-KxnH5dU9mDoveBFzoif:b21hxrq) ([raw file](https://github.com/robotlocomotion/drake/blob/37dc53ccdce36f02fca115fb69d23049ad55d544/drake/common/hash.h#L82)):*
> ```objc
>     HashAlgorithm& hasher, const T& item) noexcept {
>   // Hashing a NaN makes no since, since they cannot compare as equal.
>   DRAKE_ASSERT(!std::isnan(item));
> ```

FYI, need `#include <cmath>` for `std::isnan` (and CI fails because of this).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
I like it (speaking only to the parts that I reviewed).  I'll go and refresh/reboot #6937, too.  (As I recall, the last change was to simply defer to whatever is in  `common/hash.h`, so it should be all set, except for the new layout/repository/etc.)

---

Reviewed 1 of 27 files at r1, 3 of 9 files at r3.
Review status: 20 of 27 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/hash.h, line 81 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxnROXslijg3iX8_FyQ:-KxnROXt5SO496Mju8_B:b-48f7mm) ([raw file](https://github.com/robotlocomotion/drake/blob/37dc53ccdce36f02fca115fb69d23049ad55d544/drake/common/hash.h#L81)):*
> ```objc
>     // NOLINTNEXTLINE(runtime/references)
>     HashAlgorithm& hasher, const T& item) noexcept {
>   // Hashing a NaN makes no since, since they cannot compare as equal.
> ```

Typo:  "no since" --> "no sense"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-KxnWadIhS4OfiOwk-GT:b-b3h1yj)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
> I'll go and refresh/reboot #6937, ...

This PR already has a corresponding styleguide change at RobotLocomotion/styleguide#3.

maddog-tri commented:
Even better!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-KxneHN_BUMa_EWDMhIZ:b9flqkw)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 20 of 27 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/hash.h, line 81 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-KxnROXslijg3iX8_FyQ:-KxoM2EYiwrMii6aMpwB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/37dc53ccdce36f02fca115fb69d23049ad55d544/drake/common/hash.h#L81)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Typo:  "no since" --> "no sense"?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 8 of 9 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
+@seancurtis-tri  (likewise adding self so it's on my panel)

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-Kxrpbj6GVEJ0_sKIaRM:b-9o70m4)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-xenial-clang-cmake-experimental please
@drake-jenkins-bot linux-xenial-gcc-cmake-experimental-matlab please

SeanCurtis-TRI commented:
I like it as well. The only unfortunate aspect of it is an interface that requires anti-linting comments. :(  <img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-Kxxzzvc7FPe4b90xOxk:b4754it)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
> The only unfortunate aspect of it is an interface that requires anti-linting comments.

Done.

soonho-tri commented:
Reviewed 10 of 10 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
re: anti-linting comments

I'm surprised this doesn't fail lint; passing non-const references get flagged, yes? Why doesn't this?

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-Ky2_CjSyMTiu6byScoZ:b-e7r5tv)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[WORKSPACE, line 101 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-Ky2_fm2YZ__nYt5uJBB:-Ky2_fm2YZ__nYt5uJBC:bk31ynr) ([raw file](https://github.com/robotlocomotion/drake/blob/8df09a2350eebff3ed04e704e45f953fb54a68a5/WORKSPACE#L101)):*
> ```Python
>     # TODO(jwnimmer-tri) Put this back to RobotLocomotion after it merges.
>     repository = "jwnimmer-tri/styleguide",
>     commit = "f9ee3f921ecda09dfdac6d886f4eda8d82dde0ee",
> ```

@seancurtis-tri , @jwnimmer-tri is using his own version in this PR: https://github.com/jwnimmer-tri/styleguide/commit/f9ee3f921ecda09dfdac6d886f4eda8d82dde0ee

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[WORKSPACE, line 101 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-Ky2_fm2YZ__nYt5uJBB:-Ky2a7mQ6wewtgFG32iw:b-205def) ([raw file](https://github.com/robotlocomotion/drake/blob/8df09a2350eebff3ed04e704e45f953fb54a68a5/WORKSPACE#L101)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

@seancurtis-tri , @jwnimmer-tri is using his own version in this PR: https://github.com/jwnimmer-tri/styleguide/commit/f9ee3f921ecda09dfdac6d886f4eda8d82dde0ee
</blockquote></details>

Aaaaaaaah.  THanks.  :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@soonho-tri @SeanCurtis-TRI I have updated the `hash.h` documentation.  PTAL.

---

Review status: 0 of 28 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-L-sGEReKFffaQ-CuAN_:b-recmlu)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
@jamiesnape , Is there any technical updates other than hoisting `drake` dir?

---

Review status: 0 of 28 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-L05JKYkEUchcaqWbG5t:b-42hkhs)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
The r7 was a rebase-only push.  If you select r7..r8, you can see the meaningful edits.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-L05JiqVoQR-a76SH6VH:b-ejpzeb)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
:+1: 

---

Reviewed 26 of 53 files at r7, 2 of 3 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357#-:-L05MVl0if9jvlrfeVIt:b-8fx2u7)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I've rebased and updated `styleguide` to point to the newest master (well, gh-pages) branch of the RobotLocomotion repository.  I plan to merge once CI passes.

soonho-tri commented:
Reviewed 2 of 2 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7357)*
<!-- Sent from Reviewable.io -->
